src/ansi-c/c_typecheck_expr.cpp

```diff
@@ -316,10 +316,10 @@ void c_typecheck_baset::typecheck_expr_main(exprt &expr)
       }
     }
 
-    if(has_subexpr(where, ID_side_effect))
+    if(expr.id() == ID_lambda && has_subexpr(where, ID_side_effect))
     {
       error().source_location = expr.source_location();
-      error() << "quantifier must not contain side effects" << eom;
+      error() << "lambda expressions must not contain side effects" << eom;
       throw 0;
     }
```

Comment:
I suggest to do a bit more work here: let's also catch function calls, which should be doable by using
````c++
if(has_subexpr(where, [&](const exprt &subexpr) {
       return can_cast_expr<side_effect_exprt>(subexpr) &&
              can_cast_expr<side_effect_expr_function_callt>(subexpr);
    }))
{
  error().source_location = expr.source_location();
  error() << "quantifier must not contain function calls" << eom;
  throw 0;
}
````